source/Calamari.Tests/AppServiceBehaviorFixture.cs

```diff
@@ -17,279 +18,466 @@
-                context.Variables["Octopus.Action.Azure.WebAppName"] = javaSite.Name;
-                context.Variables[PackageVariables.SubstituteInFilesTargets] = "test.jsp";
-            }).Execute();
+            resourceGroupName = Guid.NewGuid().ToString();
+
+            clientId = ExternalVariables.Get(ExternalVariable.AzureSubscriptionClientId);
+            clientSecret = ExternalVariables.Get(ExternalVariable.AzureSubscriptionPassword);
+            tenantId = ExternalVariables.Get(ExternalVariable.AzureSubscriptionTenantId);
+            subscriptionId = ExternalVariables.Get(ExternalVariable.AzureSubscriptionId);
+            resourceGroupLocation = Environment.GetEnvironmentVariable("AZURE_NEW_RESOURCE_REGION") ?? "eastus";
 
-            await AssertContent($"{javaSite.Name}.azurewebsites.net", $"Hello! {greeting}", "test.jsp");
+            authToken = await Auth.GetAuthTokenAsync(activeDirectoryEndpointBaseUri, resourceManagementEndpointBaseUri,
+                tenantId, clientId, clientSecret);
 
+            var resourcesClient = new ResourcesManagementClient(subscriptionId,
+                new ClientSecretCredential(tenantId, clientId, clientSecret));
+
+            resourceGroupClient = resourcesClient.ResourceGroups;
+
+            var resourceGroup = new ResourceGroup(resourceGroupLocation);
+            resourceGroup = await resourceGroupClient.CreateOrUpdateAsync(resourceGroupName, resourceGroup);
+            
+            webMgmtClient = new WebSiteManagementClient(new TokenCredentials(authToken))
+            {
+                SubscriptionId = subscriptionId,
+                HttpClient = { BaseAddress = new Uri(DefaultVariables.ResourceManagementEndpoint) },
+            };
+
+            await ConfigureTestResources(resourceGroup);
         }
 
-        private void AddVariables(CommandTestBuilderContext context)
+        protected abstract Task ConfigureTestResources(ResourceGroup resourceGroup);
+        
+        [OneTimeTearDown]
+        public async Task Cleanup()
         {
-            context.Variables.Add(AccountVariables.ClientId, clientId);
-            context.Variables.Add(AccountVariables.Password, clientSecret);
-            context.Variables.Add(AccountVariables.TenantId, tenantId);
-            context.Variables.Add(AccountVariables.SubscriptionId, subscriptionId);
-            context.Variables.Add("Octopus.Action.Azure.ResourceGroupName", resourceGroupName);
-            context.Variables.Add("Octopus.Action.Azure.WebAppName", site.Name);
-            context.Variables.Add("Greeting", greeting);
-            context.Variables.Add(KnownVariables.Package.EnabledFeatures, KnownVariables.Features.SubstituteInFiles);
-            context.Variables.Add(PackageVariables.SubstituteInFilesTargets, "index.html");
-            context.Variables.Add(SpecialVariables.Action.Azure.DeploymentType, "ZipDeploy");
+            await resourceGroupClient.StartDeleteAsync(resourceGroupName);
```

Comment:
Something I noticed when adjusting the target discovery tests was that if there are creds problems in the setup and therefore it throws out then the resouceGroupClient here is null which causes a NullReferenceException on cleanup. I'm not sure if it would end up masking the true test error or not but it was a bit disruptive when I was running the tests locally. You could add a null check here if you wanted to.

Reply:
Done, good call, thanks.